Validate number of GPUs in distributed_test.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47259 Validate number of GPUs in distributed_test.**

As described in https://github.com/pytorch/pytorch/issues/47257, not
using enough number of GPUs would result in an error.

As a result, before we call `init_process_group` in distributed_test, we
validate we have enough GPUs.

#Closes: https://github.com/pytorch/pytorch/issues/47257

Differential Revision: [D24699122](https://our.internmc.facebook.com/intern/diff/D24699122/)

pritamdamania committed:
Validate number of GPUs in distributed_test.

As described in https://github.com/pytorch/pytorch/issues/47257, not
using enough number of GPUs would result in an error.

As a result, before we call `init_process_group` in distributed_test, we
validate we have enough GPUs.

#Closes: https://github.com/pytorch/pytorch/issues/47257

Differential Revision: [D24699122](https://our.internmc.facebook.com/intern/diff/D24699122/)

[ghstack-poisoned]

xwang233 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -289,6 +289,9 @@ def _run(cls, rank, test_name, file_name):
         self.rank = rank
         self.file_name = file_name
+
+        if torch.cuda.device_count() != self.world_size:
```
Thanks for the fix. Please correct me if I'm wrong. 

`self.world_size` would take `WORLD_SIZE` environment variable, https://github.com/pytorch/pytorch/blob/78de12f58854237dc5987e10bb4ab4fb33b2d01b/torch/testing/_internal/distributed/distributed_test.py#L317-L321
which is defined here for NCCL backend https://github.com/pytorch/pytorch/blob/78de12f58854237dc5987e10bb4ab4fb33b2d01b/test/run_test.py#L259-L263

However, `self.world_size` is not equal to `torch.cuda.device_count()` when number of GPU is greater than 3.

pritamdamania87 replied:
Good catch!

pritamdamania committed:
Update on "Validate number of GPUs in distributed_test."


As described in https://github.com/pytorch/pytorch/issues/47257, not
using enough number of GPUs would result in an error.

As a result, before we call `init_process_group` in distributed_test, we
validate we have enough GPUs.

#Closes: https://github.com/pytorch/pytorch/issues/47257

Differential Revision: [D24699122](https://our.internmc.facebook.com/intern/diff/D24699122/)

[ghstack-poisoned]

xwang233 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -289,6 +289,9 @@ def _run(cls, rank, test_name, file_name):
         self.rank = rank
         self.file_name = file_name
+
+        if torch.cuda.device_count() < int(self.world_size):
```
Thanks. Is there a reason that we have to compare `device_count` with `self.world_size`? Would 
`if torch.cuda.devide_count() == 1: sys.exit(...)` or `if self.world_size == 1: sys.exit(...)` be enough?

pritamdamania87 replied:
The test can be run by manually setting the WORLD_SIZE through an environment variable. In that case, we could have scenarios like: 1) device_count = 8, world_size=4 2) device_count=4, world_size=8. The check ensures we have enough GPUs for the given world_size.

xwang233 replied:
Oops, I missed that we can set env var of `WORLD_SIZE` manually for the test. Thanks for pointing it out!

xwang233 reviewed (approved):
Thanks for the fix! I have verified that the PR resolves the failed tests.
